src/targets/gpu/mlir_conv.cpp

```diff
@@ -0,0 +1,291 @@
+        auto flt_t = op_r->inputs().at(1)->get_shape();
+        auto out_t = op_r->get_shape();
+
+        auto i64 = shape(shape::uint64_type);
+
+        std::vector<shape> expected_inputs = {
+            flt_t, flt_t, i64, i64, i64, i64, i64, i64, i64, i64, i64, i64,  i64,   inp_t,
+            inp_t, i64,   i64, i64, i64, i64, i64, i64, i64, i64, i64, i64,  out_t, out_t,
+            i64,   i64,   i64, i64, i64, i64, i64, i64, i64, i64, i64, out_t};
+
+        return migraphx::make_op("gpu::code_object",
+                                 {
+                                     {"code_object", spec->binary},
+                                     {"symbol_name", mlir_kernel_name},
+                                     {"global", spec->global_size},
+                                     {"local", spec->local_size},
+                                     {"expected_inputs", migraphx::to_value(expected_inputs)},
+                                     {"output", migraphx::to_value(out_t)},
+                                 });
+    }
+
+    void add_memref_descriptor(std::vector<instruction_ref>& refs, instruction_ref inst)
+    {
+        const size_t offset = 0;
+        auto inst_t         = inst->get_shape();
+        refs.push_back(inst);
+        refs.push_back(inst);
+        refs.push_back(get_literal(offset)); // offset
+
+        // dim sizes
+        std::for_each(inst_t.lens().begin(), inst_t.lens().end(), [this, &refs](const auto lval) {
+            refs.push_back(get_literal(lval));
+        });
+        refs.push_back(get_literal(1)); // G
+
+        // dim strides
+        std::for_each(inst_t.strides().begin(),
+                      inst_t.strides().end(),
+                      [this, &refs](const auto lval) { refs.push_back(get_literal(lval)); });
+        refs.push_back(get_literal(1)); // G
+    }
+
+    instruction_ref insert_allocation(instruction_ref ins, const shape& s)
+    {
+        return prog->insert_instruction(ins, hip_allocate{s});
+    }
+
+    void replace_conv_op(instruction_ref ins)
+    {
+        auto conv_bin = make_mlir_binary(ins);
+        if(conv_bin)
+        {
+            auto conv = make_code_object_op(ins, conv_bin);
+
+            auto inp = ins->inputs().at(0);
+            auto flt = ins->inputs().at(1);
+            auto out = insert_allocation(ins, ins->get_shape());
+
+            std::vector<instruction_ref> refs;
+            refs.reserve(3 * 13 + 1);
+            add_memref_descriptor(refs, flt);
+            add_memref_descriptor(refs, inp);
+            add_memref_descriptor(refs, out);
+            refs.push_back(out);
+
+            prog->replace_instruction(ins, conv, refs);
+        }
+    }
+
+    void apply()
+    {
+        init();
+        for(auto it = prog->begin(); it != prog->end(); it++)
+        {
+            if(it->name() == "convolution")
+            {
+                replace_conv_op(it);
+            }
+        }
+    }
+};
+
+void mlir_conv::apply(module& p) const
```

Comment:
It would be better for the parameter to be `m`(for module) instead of `p`(which would stand for `program`).